### PR TITLE
Lots and lots of updates including the new module loading system :D

### PR DESCRIPTION
**Pull request by mrmelon54**

### CODE REVIEW

File: modules/hexamaze.py [hunk 1]
```diff
@@ -187,8 +198,10 @@ def __init__(self, bomb, ident):
         self.maze_center = q, r
         self.maze_rotation = random.randint(0, 5)
         self.pawn_color = random.randint(0, 5)
-        self.solution_edge = Hexamaze.EDGES[(self.maze_rotation + self.pawn_color) % 6]
-        self.solution_directions = [(self.pawn_color + self.maze_rotation + x) % 6 for x in range(2)]
+        self.solution_edge = Hexamaze.EDGES[(
+            self.maze_rotation + self.pawn_color) % 6]
+        self.solution_directions = [
+            (self.pawn_color + self.maze_rotation + x) % 6 for x in range(2)]
```
Comment (meithecatte) [inline modules/hexamaze.py]: What makes you think that this formatting is better?

File: modules/hexamaze.py [hunk 2]
```diff
@@ -246,19 +262,24 @@ def __init__(self, bomb, ident):
...
+        if self.maze_rotation == 1:
+            q, r = q + r, -q
+        elif self.maze_rotation == 2:
+            q, r = r, -q - r
+        elif self.maze_rotation == 3:
+            q, r = -q, -r
+        elif self.maze_rotation == 4:
+            q, r = -q - r, q
+        elif self.maze_rotation == 5:
+            q, r = -r, q + r
```
Comment (meithecatte) [inline modules/hexamaze.py]: I think the previous formatting looked better.

File: modules/wireSequence.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+    COMMANDS = {
+        'cut': cmd_cut,
+        'down': cmd_down,
+        'd': cmd_down,
+        'up': cmd_up,
+        'u': cmd_up,
+    }
+
+__module_class__=WireSequence
```
Comment (meithecatte) [inline modules/wireSequence.py]: I don't think `__module_class__` is necessary anymore?

File: modules/whos_on_first.py [hunk 4]
```diff
@@ -6,7 +6,7 @@ class WhosOnFirst(modules.Module):
     manual_name = "Who\u2019s on First"
     help_text = "`{cmd} push you're` or `{cmd} press press` to push a button. The phrase must match exactly."
     module_score = 4
-    third_base = False
+    self.transform='none'
```
Comment (meithecatte) [inline modules/whos_on_first.py]: Don't use `self` here. Also, is this variable used anywhere?

File: modules/whos_on_first.py [hunk 5]
```diff
@@ -77,20 +76,22 @@ def get_svg(self, led):
...
+
     def randomize(self):
         self.display = random.choice(list(self.DISPLAY_WORDS.keys()))
         self.buttons = random.sample(random.choice(self.BUTTON_GROUPS), 6)
         self.log(f"State randomized. Stage {self.stage}. Display: {self.display}. Buttons: {' '.join(self.buttons)}")
-    
+
+    def button_replace(self,v):
+        return v
+
```
Comment (meithecatte) [inline modules/whos_on_first.py]: I like how you're moving this clearly Third Base-specific code. However, this does not belong in this commit. Besides, make the function name descriptive. How about `canonical_button_name`? Also, add a comment explaining that the function is not a no-op in subclasses.

File: modules/whos_on_first.py [hunk 6]
```diff
@@ -77,20 +76,22 @@ def get_svg(self, led):
...
     @modules.check_solve_cmd
     async def cmd_push(self, author, parts):
         if not parts:
             return await self.usage(author)
         button = ' '.join(parts).upper()
 
-        if self.third_base:
-            button = button.replace('0', 'O').replace('1', 'I')
+        button=self.button_replace(button)
```
Comment (meithecatte) [inline modules/whos_on_first.py]: As above, not in this commit. Also, add spaces around the `=`.

File: modules/button.py [hunk 7]
```diff
@@ -97,10 +99,12 @@ def get_svg(self, led):
...
                 time = self.bomb.get_time_formatted()
             expected = self.get_release_digit()
-            self.log("Releasing at {:s}, expected {:d}, player answered {:s}".format(time, expected, answer))
+            self.log("Releasing at {:s}, expected {:d}, player answered {:s}".format(
+                time, expected, answer))
             self.strip_color = None
             should_hold = self.should_hold()
-            self.log("should{:s} hold".format("n't" if not should_hold else ''))
+            self.log("should{:s} hold".format(
+                "n't" if not should_hold else ''))
```
Comment (meithecatte) [inline modules/button.py]: What makes you think this is a formatting improvement?

File: modules/complicatedWires.py [hunk 8]
```diff
@@ -5,6 +5,7 @@
 import edgework
 
 class ComplicatedWires(modules.Module):
+    identifiers = ['complicatedWires']
```
Comment (meithecatte) [inline modules/complicatedWires.py:8]: This should not be a separate commit.

File: modules/__init__.py [hunk 9]
```diff
@@ -1,38 +1,26 @@
...
-    "thirdBase": ThirdBase,
-}
+for module_file in glob(path_join(dirname(__file__), "*.py")):
+    module_name = basename(module_file)[:-3]
+    importlib.import_module('modules.' + module_name)
 
 async def cmd_modules(channel, author, parts):
-    list_ = lambda d: ', '.join(f"`{x}`" for x in d)
+    def list_(d): return ', '.join(f"`{x}`" for x in d)
     await channel.send(f"Available modules:\nVanilla: {list_(VANILLA_MODULES)}\nModded: {list_(MODDED_MODULES)}")
```
Comment (meithecatte) [inline modules/__init__.py]: Don't squash other people's commits into yours, especially if the changes are unrelated.

File: modules/switches.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        if self.switches[index] == "0":
+            position[2][1] = 348 - position[2][1]
+            position[3][1] = 348 - position[3][1]
+
+        position = map(lambda x: map(lambda y: str(y), x), position)
+
+        o = f'<circle stroke-width="2" stroke="#000" cx="{58*(index+1)}" cy="174" r="15" />'
+        o += f'<circle stroke-width="2" stroke="#000" fill="{"#0f0" if self.solution[index]=="0" else "#fff"}" cx="{58*(index+1)}" cy="87" r="10" />'
+        o += f'<circle stroke-width="2" stroke="#000" fill="{"#0f0" if self.solution[index]=="1" else "#fff"}" cx="{58*(index+1)}" cy="261" r="10" />'
+        o += f'<polygon points="{" ".join(list(map(lambda x:",".join(x),position)))}" style="fill:black;stroke:black;stroke-width:2" />'
```
Comment (meithecatte) [inline modules/switches.py]: It's not necessary to use `list()` here.

File: modules/switches.py [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+        return o
```
Comment (meithecatte) [inline modules/switches.py]: Consider using list/generator comprehension instead of `map`.

File: modules/switches.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+        self.switches = "10000"
+        while self.switches in self.invalidPositions or self.switches == self.solution:
+            self.switches = ''.join([str(random.randint(0, 1)) for i in range(5)])
+
+        self.positionNames = ["up", "down"]
+
+        self.log(f"Solution: {' '.join(map(lambda x:self.positionNames[int(x)],list(self.solution)))}")
+
+    def generateSwitch(self, index):
+        default = [[50.5, 174], [65.5, 174], [73, 234], [43, 234]]
```
Comment (meithecatte) [inline modules/switches.py]: The inner lists should be tuples.

File: modules/switches.py [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+        return (f'<svg viewBox="0 0 348 348" fill="#fff" stroke-linecap="butt" stroke-linejoin="round" stroke-miterlimit="10" xmlns:xlink="http://www.w3.org/1999/xlink">'
+                f'<path stroke="#000" stroke-width="2" d="M5 5h338v338h-338z"/>'
+                f'<circle fill="{led}" stroke="#000" cx="298" cy="40.5" r="15" stroke-width="2"/>'
+
+                f'{self.generateSwitch(0)}'
+                f'{self.generateSwitch(1)}'
+                f'{self.generateSwitch(2)}'
+                f'{self.generateSwitch(3)}'
+                f'{self.generateSwitch(4)}'
+
```
Comment (meithecatte) [inline modules/switches.py]: This can be a loop.

File: modules/switches.py [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+    module_score = 1
+
+    def __init__(self, bomb, ident):
+        super().__init__(bomb, ident)
+
+        self.invalidPositions = "11011/10100/10000/01101/01100/01000/00111/00101/00011/00001".split("/")
+
+        self.solution = "10000"
+        while self.solution in self.invalidPositions:
+            self.solution = ''.join([str(random.randint(0, 1)) for i in range(5)])
```
Comment (meithecatte) [inline modules/switches.py]: Keep these as a list of bools, not as a string.

File: modules/switches.py [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+    identifiers = ['switches']
+    display_name = "Switches"
+    manual_name = "Switches"
+    help_text = "`{cmd} flip 1 5 3 4 2 5 2` or `{cmd} flip 1534252`. Partial solutions allowed."
+    module_score = 1
+
+    def __init__(self, bomb, ident):
+        super().__init__(bomb, ident)
+
+        self.invalidPositions = "11011/10100/10000/01101/01100/01000/00111/00101/00011/00001".split("/")
```
Comment (meithecatte) [inline modules/switches.py]: This looks like a constant. Use `UPPER_SNAKE_CASE` and define it in the class itself. Also, use a list - that's clearer and takes few seconds more to write (`:s@/@", "@` in vim, for example).

Comment (meithecatte) [reply modules/switches.py]: Actually, use a set - `{...}` instead of `[...]`. The order is irrelevant and `x in a_set` is faster than `x in a_list`.

File: modules/switches.py [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+                f'{self.generateSwitch(2)}'
+                f'{self.generateSwitch(3)}'
+                f'{self.generateSwitch(4)}'
+
+                f'</svg>')
+
+    def flip_switch_by_index(self, index):
+        splitSwitches = list(self.switches)
+        splitSwitches[index] = "1" if splitSwitches[index] == "0" else "0"
+        self.switches = "".join(splitSwitches)
```
Comment (meithecatte) [inline modules/switches.py]: Sounds like a string is not the right data structure, eh?

File: modules/switches.py [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+
+        self.invalidPositions = "11011/10100/10000/01101/01100/01000/00111/00101/00011/00001".split("/")
+
+        self.solution = "10000"
+        while self.solution in self.invalidPositions:
+            self.solution = ''.join([str(random.randint(0, 1)) for i in range(5)])
+
+        self.switches = "10000"
+        while self.switches in self.invalidPositions or self.switches == self.solution:
+            self.switches = ''.join([str(random.randint(0, 1)) for i in range(5)])
```
Comment (meithecatte) [inline modules/switches.py]: The code here is quite similar to the one above. Maybe this could be a separate function? `random_switches_state(invalid_states: set)`?

**Review by meithecatte — CHANGES_REQUESTED** [review]
I just noticed I don't have any license in this repository, so I just added one. Do you agree with licensing your contributions as AGPLv3?

File: modules/wireSequence.py (hunk 3 above)
Comment (mrmelon54) [reply modules/wireSequence.py]: oops i thought i removed all of those

File: modules/whos_on_first.py (hunk 4 above)
Comment (mrmelon54) [reply modules/whos_on_first.py]: um oops and yes it is used in the get_svg function

File: modules/utils.py [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+        "u": "..-",
+        "v": "...-",
+        "w": ".--",
+        "x": "-..-",
+        "y": "-.--",
+        "z": "--..",
+    }
+    if letter in LETTER_TO_MORSE.keys():
+        return LETTER_TO_MORSE[letter]
+    return None
```
Comment (meithecatte) [inline modules/utils.py]: Don't use None for error handling, let the exception happen.

File: modules/utils.py [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+        return MORSE_TO_LETTER[morse]
+    return None
+
+# converts lots of morse code into a string
+def morseToString(morse):
+    words = morse.split(" // ")
+    letters = map(lambda x:x.split(" "),words)
+    lettersConverted = map(lambda x:map(morseToLetter,x),letters)
+    wordsConverted = map(lambda x:"".join(x),lettersConverted)
+    return " ".join(wordsConverted)
```
Comment (meithecatte) [inline modules/utils.py]: These functions aren't being used anywhere, and I'm having a hard time imagining use-cases for some of those. You can add these when you need them.

File: modules/utils.py [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+        ".-.": "r",
+        "...": "s",
+        "-": "t",
+        "..-": "u",
+        "...-": "v",
+        ".--": "w",
+        "-..-": "x",
+        "-.--": "y",
+        "--..": "z",
+    }
```
Comment (meithecatte) [inline modules/utils.py]: This could be derived in a single line from the previous dictionary. (make sure to do it outside of the function so that it's not redone every time the function is called)

**Review by meithecatte — CHANGES_REQUESTED** [review]

**Comment by meithecatte** [comment]
There's no need to make another PR when you make changes.

File: modules/__init__.py [hunk 21]
```diff
@@ -1,19 +1,20 @@
...
     module_name=inspect.getmro(module)[0].__name__
     category = VANILLA_MODULES if module_name in LIST_OF_VANILLA_MODULES else MODDED_MODULES
     for identifier in module.identifiers:
         category[identifier] = module
 
-LIST_OF_VANILLA_MODULES=['Button','ComplicatedWires','Keypad','Maze','Memory','MorseCode','Password','SimonSays','WhosOnFirst','Wires','WireSequence']
-
-VANILLA_MODULES = {}
-MODDED_MODULES = {}
+from modules.base import Module, noparts, check_solve_cmd, gif_append, gif_output
```
Comment (meithecatte) [inline modules/__init__.py]: Please keep the comment that was here before.

**Review by meithecatte — CHANGES_REQUESTED** [review]

File: modules/__init__.py (hunk 21 above)
Comment (mrmelon54) [reply modules/__init__.py]: oh yes sorry

File: modules/whosOnFirst.py [hunk 22]
```diff
@@ -0,0 +1,138 @@
+import random
```
Comment (meithecatte) [inline modules/whosOnFirst.py]: Am I seeing things, or did you add a second copy of the who's on first code?

Comment (mrmelon54) [reply modules/whosOnFirst.py]: I renamed from whos_on_first to whosOnFirst but my files glitched out and whos_on_first was created again

**Comment by mrmelon54** [comment]
> I just noticed I don't have any license in this repository, so I just added one. Do you agree with licensing your contributions as AGPLv3?

Yes

File: modules/whosOnFirst.py [hunk 23]
```diff
@@ -2,15 +2,18 @@
...
-    third_base = False
+    transform = 'none'
 
     BUTTON_GROUPS = [
-        ["READY", "FIRST", "NO", "BLANK", "NOTHING", "YES", "WHAT", "UHHH", "LEFT", "RIGHT", "MIDDLE", "OKAY", "WAIT", "PRESS"],
-        ["YOU", "YOU ARE", "YOUR", "YOU'RE", "UR", "U", "UH HUH", "UH UH", "WHAT?", "DONE", "NEXT", "HOLD", "SURE", "LIKE"],
+        ["READY", "FIRST", "NO", "BLANK", "NOTHING", "YES", "WHAT",
+            "UHHH", "LEFT", "RIGHT", "MIDDLE", "OKAY", "WAIT", "PRESS"],
+        ["YOU", "YOU ARE", "YOUR", "YOU'RE", "UR", "U", "UH HUH",
+            "UH UH", "WHAT?", "DONE", "NEXT", "HOLD", "SURE", "LIKE"],
```
Comment (meithecatte) [inline modules/whosOnFirst.py:16]: There's no need to split these across lines.

File: modules/thirdBase.py [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+
+class ThirdBase(WhosOnFirst):
+    identifiers = ['thirdBase']
+    display_name = "Third Base"
+    manual_name = "Third Base"
+    help_text = "`{cmd} push 8i99` or `{cmd} press 66i8` to push a button."
+    module_score = 6
+    transform='rotate(180 174 174)'
+
+    def button_replace(self,v):
```
Comment (meithecatte) [inline modules/thirdBase.py]: This function is never called because you didn't change the name in both places.

File: modules/whosOnFirst.py [hunk 25]
```diff
@@ -77,20 +79,22 @@ def get_svg(self, led):
...
         return svg
-    
+
     def randomize(self):
         self.display = random.choice(list(self.DISPLAY_WORDS.keys()))
         self.buttons = random.sample(random.choice(self.BUTTON_GROUPS), 6)
         self.log(f"State randomized. Stage {self.stage}. Display: {self.display}. Buttons: {' '.join(self.buttons)}")
-    
+
+    def button_replace_values(self, v):
```
Comment (meithecatte) [inline modules/whosOnFirst.py]: As I've said before, add a comment explaining why this function is a no-op. I understand why this is the case, but it doesn't mean I'll remember that forever. Also, I feel like `button_replace_values` isn't really descriptive. My first guess at what this does is "choose new values for the buttons", like what `randomize` currently does. How about calling this `preprocess_input`? I don't think that can be misunderstood.

File: modules/whosOnFirst.py [hunk 26]
```diff
@@ -77,20 +79,22 @@ def get_svg(self, led):
...
+
     @modules.check_solve_cmd
     async def cmd_push(self, author, parts):
         if not parts:
             return await self.usage(author)
         button = ' '.join(parts).upper()
 
-        if self.third_base:
-            button = button.replace('0', 'O').replace('1', 'I')
+        button = self.button_replace_values(button)
```
Comment (meithecatte) [inline modules/whosOnFirst.py]: Consider merging this into one line: `button = self.preprocess_input(' '.join(parts).upper())`

File: modules/wireSequence.py (hunk 3 above)
Comment (meithecatte) [reply modules/wireSequence.py]: Please don't fix wrong code by adding a commit at the end. It makes reviews confusing and breaks `git bisect`.

**Review by meithecatte — CHANGES_REQUESTED** [review]

File: modules/thirdBase.py (hunk 24 above)
Comment (mrmelon54) [reply modules/thirdBase.py]: this is for an outdated commit

File: modules/whosOnFirst.py (hunk 25 above)
Comment (mrmelon54) [reply modules/whosOnFirst.py]: this is for an outdated commit

File: modules/thirdBase.py (hunk 24 above)
Comment (meithecatte) [reply modules/thirdBase.py]: See other comments.

File: modules/whosOnFirst.py (hunk 26 above)
Comment (mrmelon54) [reply modules/whosOnFirst.py]: also for an outdated commit

File: modules/switches.py (hunk 10 above)
Comment (mrmelon54) [reply modules/switches.py]: oh whoops I thought I removed all of those

File: modules/switches.py [hunk 27]
```diff
@@ -0,0 +1,107 @@
...
+            if part.isdigit():
+                for digit in part:
+                    switch = int(digit)
+                    if switch not in range(1, 6):
+                        return await self.bomb.channel.send(f"{author.mention} There is no switch {switch}! There are five switches: 1-5.")
+                    input_.append(switch - 1)
+            else:
+                return await self.bomb.channel.send(f"{author.mention} No such switch: `{part}`")
+
+        while input_ and not self.solved:
```
Comment (meithecatte) [inline modules/switches.py:94]: As far as I can see, this `self.solved` condition can never be triggered, because you only set it while calling `handle_solve` below the loop.

File: modules/thirdBase.py [hunk 28]
```diff
@@ -0,0 +1,55 @@
...
+import modules
+from modules.whosOnFirst import WhosOnFirst
+
+class ThirdBase(WhosOnFirst):
+    identifiers = ['thirdBase']
+    display_name = "Third Base"
+    manual_name = "Third Base"
+    help_text = "`{cmd} push 8i99` or `{cmd} press 66i8` to push a button."
+    module_score = 6
+    transform='rotate(180 174 174)'
```
Comment (meithecatte) [inline modules/thirdBase.py]:
```suggestion
    transform = 'rotate(180 174 174)'
```

File: modules/utils.py [hunk 29]
```diff
@@ -0,0 +1,39 @@
...
+
+# turns a letter into the morse code equivalent
+# or `None` for an invalid input
+def letterToMorse(letter):
+    return LETTER_TO_MORSE[letter]
+
+# turns a morse code string  equivalent
+# or `None` for an invalid input
+def morseToLetter(morse):
+    return MORSE_TO_LETTER[morse]
```
Comment (meithecatte) [inline modules/utils.py]: Is this actually used anywhere?

File: modules/switches.py (hunk 27 above)
Comment (mrmelon54) [reply modules/switches.py:94]: I some how just completely forgot to add that condition

File: modules/whosOnFirst.py [hunk 30]
```diff
@@ -91,8 +91,8 @@ def randomize(self):
             return await self.usage(author)
         button = ' '.join(parts).upper()
 
-        if self.third_base:
-            button = button.replace('0', 'O').replace('1', 'I')
+        if hasattr(self,"canonical_button_name"):
+            button = self.canonical_button_name(button)
```
Comment (meithecatte) [inline modules/whosOnFirst.py]: There's no need for reflection here. Implement `canonical_button_name` that just returns its argument, and explain in a comment that Third Base, being a subclass, overwrites the method.

File: modules/wireSequence.py [hunk 31]
```diff
@@ -0,0 +1,205 @@
+import random
+import itertools
+import enum
+import modules
+import edgework
```
Comment (meithecatte) [inline modules/wireSequence.py]: There's no need to import itertools nor edgework here.

File: modules/switches.py [hunk 32]
```diff
@@ -0,0 +1,107 @@
...
+        while not self.is_valid_state(self.solution):
+            self.solution = self.gen_rand()
+
+        self.switches = self.gen_rand()
+        while True:
+            if self.is_valid_state(self.switches) and self.switches != self.solution:
+                break
+            self.switches = self.gen_rand()
+
+        init_str=" ".join(map(lambda x:("UP","DOWN")[x],self.switches))
```
Comment (meithecatte) [inline modules/switches.py]: I feel like this should be refactored into a separate function: `state_as_string`. You could then use it to add some logging after each switch being flipped.

File: modules/switches.py [hunk 33]
```diff
@@ -0,0 +1,107 @@
...
+        sltn_str=" ".join(map(lambda x:("UP","DOWN")[x],self.solution))
+        self.log(f"Initial position: {init_str}")
+        self.log(f"Solution: {sltn_str}")
+
+    def generateSwitch(self, index):
+        position = list(map(lambda x: [x[0] + 58 * index, x[1]], self.positions_for_svg_generation))
```
Comment (meithecatte) [inline modules/switches.py]:
```suggestion
        positions = [[x[0] + 58 * index, x[1]] for x in self.positions_for_svg_generation]
```

File: modules/switches.py [hunk 34]
```diff
@@ -0,0 +1,107 @@
...
+        if self.switches[index] == self.D.UP:
+            position[2][1] = 348 - position[2][1]
+            position[3][1] = 348 - position[3][1]
+
+        position = map(lambda x: map(lambda y: str(y), x), position)
```
Comment (meithecatte) [inline modules/switches.py]:
```suggestion
        positions = " ".join(f"{pos[0]},{pos[1]}" for pos in positions)
```

File: modules/switches.py [hunk 35]
```diff
@@ -0,0 +1,107 @@
+import random
+import modules
+import enum
+
+class Switches(modules.Module):
+    class D():
+        UP=False
+        DOWN=True
```
Comment (meithecatte) [inline modules/switches.py]: Formatting.

File: modules/switches.py [hunk 36]
```diff
@@ -0,0 +1,107 @@
...
+    identifiers = ['switches']
+    display_name = "Switches"
+    manual_name = "Switches"
+    help_text = "`{cmd} flip 1 5 3 4 2 5 2` or `{cmd} flip 1534252`. Partial solutions allowed."
+    module_score = 4
+
+    # each number is a switch in the up state
+    invalid_positions = ["3","245","2345","14","145","1345","12","124","123","1234"]
+    positions_for_svg_generation = [[50.5, 174], [65.5, 174], [73, 234], [43, 234]]
```
Comment (meithecatte) [inline modules/switches.py]: I feel that a tuple fits better for representing coordinates:

```suggestion
    switch_polygon = [(50.5, 174), (65.5, 174), (73, 234), (43, 234)]
```

File: modules/switches.py [hunk 37]
```diff
@@ -0,0 +1,107 @@
...
+        UP=False
+        DOWN=True
+    identifiers = ['switches']
+    display_name = "Switches"
+    manual_name = "Switches"
+    help_text = "`{cmd} flip 1 5 3 4 2 5 2` or `{cmd} flip 1534252`. Partial solutions allowed."
+    module_score = 4
+
+    # each number is a switch in the up state
+    invalid_positions = ["3","245","2345","14","145","1345","12","124","123","1234"]
```
Comment (meithecatte) [inline modules/switches.py]:
```suggestion
    invalid_positions = {"3","245","2345","14","145","1345","12","124","123","1234"}
```

File: modules/wireSequence.py (hunk 31 above)
Comment (meithecatte) [reply modules/wireSequence.py]: Same goes for `wires.py`, actually.

File: modules/whosOnFirst.py (hunk 30 above)
Comment (mrmelon54) [reply modules/whosOnFirst.py]: ok

File: modules/switches.py (hunk 32 above)
Comment (mrmelon54) [reply modules/switches.py]: yeah that might be helpful

File: modules/switches.py (hunk 35 above)
Comment (mrmelon54) [reply modules/switches.py]: um i always forget

File: modules/__init__.py [hunk 38]
```diff
@@ -1,12 +1,16 @@
...
+
+LIST_OF_VANILLA_MODULES=['Button','ComplicatedWires','Keypad','Maze','Memory','MorseCode','Password','SimonSays','WhosOnFirst','Wires','WireSequence']
 
 VANILLA_MODULES = {}
 MODDED_MODULES = {}
 
 def register_module(module):
-    category = VANILLA_MODULES if module.vanilla else MODDED_MODULES
+    module_name=inspect.getmro(module)[0].__name__
+    category = VANILLA_MODULES if module_name in LIST_OF_VANILLA_MODULES else MODDED_MODULES
```
Comment (meithecatte) [inline modules/__init__.py:13]: Can you remind me what this change is for?

File: modules/switches.py [hunk 39]
```diff
@@ -0,0 +1,109 @@
+import random
+import modules
+import enum
+
+class Switches(modules.Module):
+    class D():
```
Comment (meithecatte) [inline modules/switches.py]:
```suggestion
    class D:
```

File: modules/switches.py [hunk 40]
```diff
@@ -0,0 +1,109 @@
...
+    # lists indexes of switches in the up position
+    def up_state_positions(self, switch_list):
+        return "".join([str(i+1) for i, val in enumerate(switch_list) if val == self.D.UP])
+
+    # checks if the state is valid
+    def is_valid_state(self, switch_state):
+        return not self.up_state_positions(switch_state) in self.invalid_positions
+
+    def state_as_string(self,state):
+        return " ".join(map(lambda x:("UP", "DOWN")[x], state))
```
Comment (meithecatte) [inline modules/switches.py]:
```suggestion
        return " ".join("UP" if switch == self.D.UP else "DOWN" for switch in state)
```

File: modules/switches.py [hunk 41]
```diff
@@ -0,0 +1,109 @@
...
+
+    # lists indexes of switches in the up position
+    def up_state_positions(self, switch_list):
+        return "".join([str(i+1) for i, val in enumerate(switch_list) if val == self.D.UP])
+
+    # checks if the state is valid
+    def is_valid_state(self, switch_state):
+        return not self.up_state_positions(switch_state) in self.invalid_positions
+
+    def state_as_string(self,state):
```
Comment (meithecatte) [inline modules/switches.py]:
```suggestion
    def state_as_string(self, state):
```

File: modules/switches.py [hunk 42]
```diff
@@ -0,0 +1,109 @@
...
+    help_text = "`{cmd} flip 1 5 3 4 2 5 2` or `{cmd} flip 1534252`. Partial solutions allowed."
+    module_score = 4
+
+    # each number is a switch in the up state
+    invalid_positions = {"3","245","2345","14","145","1345","12","124","123","1234"}
+    switch_polygon = [(50.5, 174), (65.5, 174), (73, 234), (43, 234)]
+
+    # lists indexes of switches in the up position
+    def up_state_positions(self, switch_list):
+        return "".join([str(i+1) for i, val in enumerate(switch_list) if val == self.D.UP])
```
Comment (meithecatte) [inline modules/switches.py]:
```suggestion
        return "".join(str(i + 1) for i, val in enumerate(switch_list) if val == self.D.UP)
```

File: modules/thirdBase.py [hunk 43]
```diff
@@ -0,0 +1,55 @@
...
+
+class ThirdBase(WhosOnFirst):
+    identifiers = ['thirdBase']
+    display_name = "Third Base"
+    manual_name = "Third Base"
+    help_text = "`{cmd} push 8i99` or `{cmd} press 66i8` to push a button."
+    module_score = 6
+    transform = 'rotate(180 174 174)'
+
+    def canonical_button_name(self,v):
```
Comment (meithecatte) [inline modules/thirdBase.py]:
```suggestion
    def canonical_button_name(self, v):
```

File: modules/utils.py [hunk 44]
```diff
@@ -0,0 +1,29 @@
...
+    "s": "...",
+    "t": "-",
+    "u": "..-",
+    "v": "...-",
+    "w": ".--",
+    "x": "-..-",
+    "y": "-.--",
+    "z": "--..",
+}
+MORSE_TO_LETTER = dict([(value, key) for key, value in LETTER_TO_MORSE.items()])
```
Comment (meithecatte) [inline modules/utils.py:29]:
```suggestion
MORSE_TO_LETTER = {code: letter for letter, code in LETTER_TO_MORSE.items()}
```

File: modules/whosOnFirst.py [hunk 45]
```diff
@@ -85,14 +85,16 @@ def randomize(self):
         self.buttons = random.sample(random.choice(self.BUTTON_GROUPS), 6)
         self.log(f"State randomized. Stage {self.stage}. Display: {self.display}. Buttons: {' '.join(self.buttons)}")
 
+    # method is overwritten when used in third base
+    def canonical_button_name(self,v):
```
Comment (meithecatte) [inline modules/whosOnFirst.py:89]:
```suggestion
    def canonical_button_name(self, v):
```

**Review by meithecatte — COMMENTED** [review]
This is getting better and better! Apart from some minor improvements (that I'm putting here as suggestions so that you can learn from them), I have two concerns:
 - The branch is conflicting with master, so a rebase will probably be necessary. While you're at it, could you squash these into two commits: the refactoring, and then everything that touches `switches.py`?
 - The original module [makes sure not to generate configurations solvable with one move](https://github.com/bfdev/briansktanemods/blob/SwitchesModule/Assets/SwitchesModule/Scripts/SwitchConfiguration.cs#L32-L45). It would be nice to replicate this here.

File: modules/__init__.py (hunk 38 above)
Comment (mrmelon54) [reply modules/__init__.py:13]: Prevents every non-vanilla module that uses a vanilla module as a super class from having `vanilla = False` line

Comment (meithecatte) [reply modules/__init__.py:13]: I thought about it, and the issue stems from a misuse of OOP. ThirdBase shouldn't be a subclass of WhosOnFirst -- they should have a common superclass instead. I'll do the change in a moment.

**Comment by meithecatte** [comment]
We've merged this, right?

**Comment by mrmelon54** [comment]
i have no idea 😹

**Comment by meithecatte** [comment]
Well, switches are there, so I would guess that the rest is too. I don't know why this stayed open after I've merged this, maybe I did it manually and forgot to close.

**Comment by mrmelon54** [comment]
Yeah maybe
